### ISSUE TITLE
[BUG] Question / Possible bug - observation data changing when loaded into TransitionMiniBatch()

### ISSUE THREAD

**Issue by emrul** [takuseno/d3rlpy#264]
This is more a question than a bug report, to help my understanding of how things work.

**Describe the bug**
I was stepping through the `fit` code to understand how it works in PyCharm and I noticed something odd. My observation space is a 2d array and almost all values are binary 0/1 except very few.

When I inspect the `batch` object and look at `next_observations` or `observations` I see that some values I expect to be 0 or 1 are in fact different values:
<img width="1369" alt="ml3023py_–___dev_trading_ml3023py_lib_python3_10_site-packages_d3rlpy_base_py" src="https://user-images.githubusercontent.com/674888/220240411-d0a042d1-74e4-4712-9530-19a875d027c6.png">

I thought maybe the problem was in the way I generated the dataset but if I inspect the `dataset` property I see that the values are correct:
<img width="1379" alt="ml3023py_–___dev_trading_ml3023py_lib_python3_10_site-packages_d3rlpy_base_py" src="https://user-images.githubusercontent.com/674888/220240534-d0df58eb-6018-4044-abd5-0f0f55cdb7d5.png">

Why does the `batch` observation values not match the observations in my dataset? I thought maybe there was a scaler being applied to my observations but `self.scaler` is `None`

Thank you and apologies in advance for such a basic question.

I did some more step-debugging and the change happens in [here](https://github.com/takuseno/d3rlpy/blob/32e845eb2feb0763275501ed8ee677c62ae19ca5/d3rlpy/iterators/base.py#L53-L60) - `transitions` contains just 0s and 1s as I expect but when I inspect `next_observations` or `observations`
 in `batch` some values are altered.

**Comment by takuseno** [comment]
@emrul Hi, thank you for reporting this. I guess this is because your observation numpy arrays are not contiguous and `memcpy` does not work properly with it. To construct a mini-batch, `memcpy` is used to reduce the overhead.
https://github.com/takuseno/d3rlpy/blob/32e845eb2feb0763275501ed8ee677c62ae19ca5/d3rlpy/dataset.pyx#L1260

Please make sure that your numpy arrays are contiguous.
https://numpy.org/doc/stable/reference/generated/numpy.ascontiguousarray.html

**Comment by emrul** [comment]
Hi @takuseno - I did not think the issue originated in my observations. To be sure, I updated my dataset creation code (made using the guide [here](https://d3rlpy.readthedocs.io/en/v1.1.1/tutorials/create_your_dataset.html)) to include `ascontiguousarray` and re-saved my h5 file and then retested - the same issue occurs.  I verified my observation using `observation.flags['C_CONTIGUOUS'] == true` (from the np docs you linked to)

To be extra sure, I placed a breakpoint in `iterators/base.py` and checked that the observations are contiguous there too in my debugger.

I don't think therefore contiguous arrays are the problem?
<img width="1399" alt="ml3023py_–_iterators_base_py" src="https://user-images.githubusercontent.com/674888/220373323-2ccba16a-f951-4ab6-8527-84f377da5209.png">

**Comment by takuseno** [comment]
I see. Can you share the minimum example to reproduce your issue? In unit tests, it's been tested work properly.

**Comment by emrul** [comment]
Yes, certainly - and thank you for looking into this so fast. I hope the following code is sufficiently minimal to reproduce the issue:

```
import d3rlpy
from d3rlpy.dataset import MDPDataset
from d3rlpy.algos import DQN
from d3rlpy.iterators import RoundIterator
from d3rlpy.models.encoders import EncoderFactory
from torch import nn
import numpy as np

def create_dummy_dataset():
    observations = []
    actions = []
    rewards = []
    terminals = []
    # Create a 1 state transitions where observation is a 2d array shape (380x145)
    # For each observation, [0,0] is a continuous value and column at index 6 is a continuous value
    # All other values are 0/1
    for i in range(100):
        obs = np.random.randint(0, 2, size=(100, 100)).astype(dtype=np.float32)
        obs[0,0] = np.random.rand()
        obs[:,6] = np.random.rand()
        observations.append(obs)
        assert obs.flags['C_CONTIGUOUS'], 'Obs must be contiguous'
        actions.append(np.random.randint(0, 3))
        rewards.append(0)
        terminals.append(np.random.choice([True, False]))

    dummy_dataset = MDPDataset(
        observations=np.ascontiguousarray(observations),
        actions=np.ascontiguousarray(actions),
        rewards=np.ascontiguousarray(rewards),
        terminals=np.ascontiguousarray(terminals),
    )
    return dummy_dataset

class Dummy2dModule(nn.Module):
    def __init__(self, feature_size=32):
        super(Dummy2dModule, self).__init__()
        self.feature_size = feature_size
        self.model = nn.Sequential(nn.Flatten(), nn.LazyLinear(out_features=self.feature_size))

    def get_feature_size(self):
        return self.feature_size

    def forward(self, X):
        return self.model(X)

class Dummy2dEncoder(EncoderFactory):
    TYPE = "Dummy2dObsEncoder"

    def __init__(self, feature_size=32):
        self.feature_size = feature_size

    def create(self, observation_shape):
        return Dummy2dModule(feature_size=self.feature_size)

    def get_params(self, deep=False):
        return {"feature_size": self.feature_size}

def check_data_valid(arr):
    # Mask out the values we expect to be continuous
    arr[0, 0] = 0
    arr[:, 6] = 1
    # Test everything else is 0/1
    is_all_binary = np.array_equal(arr, arr.astype(bool))
    return is_all_binary

# dataset = MDPDataset.load('repro.h5')
dataset = create_dummy_dataset()

dqn = DQN(use_gpu=False, encoder_factory=Dummy2dEncoder())

transitions = []
if isinstance(dataset, MDPDataset):
    for episode in dataset.episodes:
        transitions += episode.transitions

iterator = RoundIterator(
    transitions,
    batch_size=dqn._batch_size,
    n_steps=dqn._n_steps,
    gamma=dqn._gamma,
    n_frames=dqn._n_frames,
    real_ratio=dqn._real_ratio,
    generated_maxlen=dqn._generated_maxlen,
    shuffle=False,
)

batch = next(iterator)

# transitions[...].observation - data matches what is in the dataset
assert check_data_valid(transitions[0].observation), f'Transitions data contains non-0/1'

# batch.observations - data no longer matches, there are continuous values outside of [0,0] and [:,6]
assert check_data_valid(batch.observations[0]), f'Batch data contains non-0/1'


```

Output: `AssertionError: Batch data contains non-0/1`

**Comment by takuseno** [comment]
Currently, d3rlpy only supports 1d and 3d observation shapes. In the next update, it'll be naturally supported. But, for now, you need to flatten the observations and reshape them in `Encoder` parts.

**Comment by takuseno** [comment]
OKay, I just noticed that you sent a PR. Thanks! I'll take a look.

**Comment by emrul** [comment]
> OKay, I just noticed that you sent a PR. Thanks! I'll take a look.

Amazing, yes its quite simple to fix this. I don't think supporting only 1d and 3d observations is a limitation that makes any sense so hopefully the PR gets accepted - I'm using the version I contributed for my own training at the moment.

**Comment by takuseno** [comment]
I think this is fixed by #265 . Thank you for your contribution! @emrul
